### PR TITLE
Rewriting splitting to be more precise

### PR DESCRIPTION
**Pull request by turian**
PLEASE REVIEW CAREFULLY.

Particularly for small data set, we want accurate splitting. So instead of using which_set, which sucks for small corpora, we directly sample the exact number of files we want.

I remove split_key in this PR, which is a mistake. In https://github.com/neuralaudio/hear-preprocess/pull/8 I make sure that the splitting uses a split_key, but the exact number we want.

### CODE REVIEW

File: heareval/tasks/pipeline.py [hunk 1]
```diff
@@ -186,25 +187,6 @@ def slugify_file_name(relative_path: str):
         slug_text = slug_text.replace("-", "_negative_")
         return f"{slugify(slug_text)}"
```
Comment (turian) [inline heareval/tasks/pipeline.py:189]: I accidentally remove split_key in this PR. I add it back in https://github.com/neuralaudio/hear-preprocess/pull/8

File: heareval/tasks/pipeline.py [hunk 2]
```diff
@@ -286,37 +271,65 @@ def split_train_test_val(self, metadata: pd.DataFrame):
...
-                )
+        train_percentage: float
+        valid_percentage: float
+        test_percentage: float
+
+        # If we want a 60/20/20 split, but we already have test
+        # then we want to do a 75/25/0 split so that train is still 3x validation
+        if splits_to_sample == set():
+            return metadata
+        elif splits_to_sample == set("valid"):
```
Comment (jorshi) [inline heareval/tasks/pipeline.py:282]: This is incorret:
```
set("valid") = {'v', 'a', 'l', 'i', 'd'}
```
You want:
```
set(["valid"])

File: heareval/tasks/pipeline.py [hunk 3]
```diff
@@ -286,37 +271,65 @@ def split_train_test_val(self, metadata: pd.DataFrame):
...
+            tot = TRAIN_PERCENTAGE + TEST_PERCENTAGE
+            train_percentage = (
+                TRAIN_PERCENTAGE + TRAIN_PERCENTAGE * VALIDATION_PERCENTAGE / tot
             )
-        )
+            valid_percentage = 0
```
Comment (jorshi) [inline heareval/tasks/pipeline.py:287]: I'm confused, if the split we want to sample is "valid", then why do we set the percentage to zero?

File: heareval/tasks/pipeline.py [hunk 4]
```diff
@@ -286,37 +271,65 @@ def split_train_test_val(self, metadata: pd.DataFrame):
...
+            test_percentage = 0
+        else:
+            train_percentage = TRAIN_PERCENTAGE
+            valid_percentage = VALIDATION_PERCENTAGE
+            test_percentage = TEST_PERCENTAGE
+        assert (
+            train_percentage + valid_percentage + test_percentage == 100
+        ), f"{train_percentage + valid_percentage + test_percentage} != 100"
+
+        relpaths = metadata["relpath"].unique()
```
Comment (jorshi) [inline heareval/tasks/pipeline.py:308]: This looks like it samples from everything - Don't we always sample just from the train set only? If a test set is already present, then we want to make sure we leave that alone right?

File: heareval/tasks/pipeline.py [hunk 5]
```diff
@@ -286,37 +271,65 @@ def split_train_test_val(self, metadata: pd.DataFrame):
...
+        train_percentage: float
+        valid_percentage: float
+        test_percentage: float
+
+        # If we want a 60/20/20 split, but we already have test
+        # then we want to do a 75/25/0 split so that train is still 3x validation
+        if splits_to_sample == set():
+            return metadata
+        elif splits_to_sample == set("valid"):
+            tot = TRAIN_PERCENTAGE + TEST_PERCENTAGE
```
Comment (jorshi) [inline heareval/tasks/pipeline.py:283]: The tot whether your sampling just "valid" or just "test" is the same -- is this not supposed to be TRAIN_PERCENTAGE + VALIDATION_PERCENTAGE?

File: heareval/tasks/pipeline.py (hunk 3 above)
Comment (jorshi) [reply heareval/tasks/pipeline.py:287]: I find this calculation of percentages a bit hard to follow. Can this be simplified?

When you need to sample validation from training:
```
tot  = TRAIN_PERCENTAGE + VALIDATION_PERCENTAGE
train_percentage = TRAIN_PERCENTAGE + TRAIN_PERCENTAGE * VALIDATION_PERCENTAGE / tot
validation_percentage = 100.0 - train_percentage
test_percentage = 0.0
```

Comment (jorshi) [reply heareval/tasks/pipeline.py:287]: Unless I'm missing something.

**Review by jorshi — COMMENTED** [review]
I love that we are switching to this style of split subsampling. I don't think this code is working correctly when we need to sample just valid or just test though.

File: heareval/tasks/pipeline.py (hunk 2 above)
Comment (turian) [reply heareval/tasks/pipeline.py:282]: wow oops. added an assert too

File: heareval/tasks/pipeline.py (hunk 3 above)
Comment (turian) [reply heareval/tasks/pipeline.py:287]: "I'm confused, if the split we want to sample is "valid", then why do we set the percentage to zero?" My mistake, I'm fixing

Comment (turian) [reply heareval/tasks/pipeline.py]: @jorshi check now?

File: heareval/tasks/pipeline.py (hunk 4 above)
Comment (turian) [reply heareval/tasks/pipeline.py]: epic facepalm

Comment (turian) [reply heareval/tasks/pipeline.py]: check now?

File: heareval/tasks/pipeline.py (hunk 5 above)
Comment (turian) [reply heareval/tasks/pipeline.py]: Yeah look now

**Review by jorshi — APPROVED** [review]
Nice changes look good
